src/cpu/flexcpu/inflight_inst.cc

```diff
@@ -109,6 +112,78 @@ InflightInst::addDependency(shared_ptr<InflightInst> parent)
     });
 }
 
+void
+InflightInst::addEffAddrCalculatedCallback(function<void()> callback)
+{
+    effAddrCalculatedCallbacks.push_back(callback);
+}
+
+void
+InflightInst::addMemReadyCallback(function<void()> callback)
+{
+    memReadyCallbacks.push_back(callback);
+}
+
+void
+InflightInst::addMemCommitDependency(shared_ptr<InflightInst> parent)
+{
+    if (parent->isSquashed() || parent->isCommitted()) return;
+
+    ++remainingMemDependencies;
+
+    weak_ptr<InflightInst> weak_this = shared_from_this();
+    parent->addCommitCallback([weak_this]() {
+        shared_ptr<InflightInst> inst_ptr = weak_this.lock();
+        if (!inst_ptr) return;
+
+        --inst_ptr->remainingMemDependencies;
+
+        if (!inst_ptr->remainingMemDependencies) {
+            inst_ptr->notifyMemReady();
+        }
+    });
+}
+
+void
+InflightInst::addMemDependency(shared_ptr<InflightInst> parent)
+{
+    if (parent->isSquashed() || parent->isComplete()) return;
+
+    ++remainingMemDependencies;
+
+    weak_ptr<InflightInst> weak_this = shared_from_this();
+    parent->addCompletionCallback([weak_this]() {
+        shared_ptr<InflightInst> inst_ptr = weak_this.lock();
+        if (!inst_ptr) return;
+
+        --inst_ptr->remainingMemDependencies;
```

Comment:
General comment: This code (subtract dependencies and test for notify) is repeated multiple times. I think it should be a member function of InflightInst.

In fact, looking at this code... maybe we shouldn't be using so many lambdas here. It might make more sense to "hard code" the behavior for commit/mem/complete and use lists of instructions instead of lists of functions. I'm mostly saying this so we have a record of it for later.

Reply:
Well it has to do with the fact that it's different stages that determine one stage's dependencies. If you wanted to hard code the behavior, it'd still have to be something along the lines of "decrementAndCheckMemDependencies()" which would be called for every child of the parent, instead of through the single for loop in notifyX().

Reply:
The lambda allows us the capture a reference to "this" and store it in the parent.